pipelines: Update pipeline concept docs

dandawg opened the pull request:
<!-- Add the component name to the PR's title. Example: pipelines: Fixed broken link in Getting Started with Kubeflow Pipelines -->

The [concept page for pipelines](https://www.kubeflow.org/docs/components/pipelines/concepts/pipeline/) needed a major update to make it more relevant for ML engineers. I updated the page to include content on what pipelines are, why they are important to ML engineering, and basic pipeline construction. I also made a concerted effort to tie this page to other related resources. As part of this effort, I did some restructuring of the documentation. Below is a summary of the changes I made:

- Updated the pipeline.md concept doc
- Repurposed the current pipeline-root.md concept doc. This doc had a lot of "How to Guide" content (but not as much conceptual info), so I moved that content to a new pipeline-root.md file under the user guide section. 
- Created a new pipeline root doc focussing on the concept
- I surfaced the IR YAML concept by creating a dedicated "IR YAML" concept page. This content was originally the last section in the [Compile a Pipeline guide](https://www.kubeflow.org/docs/components/pipelines/user-guides/core-functions/compile-a-pipeline/#ir-yaml), where it wasn't very accessible. Since IR YAML is talked about in both pipeline and component concepts, it is unique to KFP, and it's touched by the user, I felt this merited its own concept page.

I added a couple of (non visible comment) TODOs where I saw future work might be needed. I deliberately did not update the "Next Steps" sections of docs. While these need updating to improve the user experience, I feel like that is better left for a future PR.

**Checklist:**
- [ ] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [ ] Ensure you follow best practices from our guide. [Contributing](https://github.com/kubeflow/website/blob/master/content/en/docs/about/contributing.md). 
- [] You have included screenshots when changing the website style or adding a new page.


**Description of your changes:**


### Issue

<!--
 If this pull request resolves an open issue, please link the issue in the PR
 description so it will automatically close when the PR is merged.

 See the GitHub documentation for more details and other options:

 https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

Closes: #

<!--Additional Information:-->
### Labels
<!-- Please include labels below by uncommenting them to help us better review PRs -->

<!-- /area central-dashboard -->
<!-- /area katib -->
<!-- /area kserve -->
<!-- /area model-registry -->
<!-- /area notebooks -->
<!-- /area pipelines -->
<!-- /area spark-operator -->
<!-- /area trainer -->
<!-- /area gsoc -->
<!-- /area website -->
<!-- /area community -->
---

dandawg commented:
This is ready to review. It looks like a lot, but much of it is just updating links, and moving some content to better locations.

The two major additions are the `pipeline.md` and `pipeline-root.md` file updates in the concept folder. Let me know if you have any questions.

rimolive commented:
I know this is not part of your PR, but since you started changing the pipeline root page could you fix that wrong link markdown?

![Screenshot from 2025-04-02 14-01-25](https://github.com/user-attachments/assets/1006daf3-709e-4be9-93f2-31c908e99db6)

rimolive commented on content/en/docs/components/pipelines/concepts/pipeline.md:
```diff
@@ -5,20 +5,126 @@ weight = 10
+- **[component][component] tasks** - define the individual pipeline steps that will run in containers
+
+### 3. Control Logic
+- **flow logic** - sequential, parallel (see [Control Flow][Control Flow])
```
Control Flow link is broken

dandawg replied:
fixed

rimolive commented on content/en/docs/components/pipelines/concepts/pipeline.md:
```diff
@@ -5,20 +5,126 @@ weight = 10
+- **exit handling** - handles exit tasks, even if upstream tasks fail (see [Control Flow/Exit Handling][Exit Handling])
+
+### 4. Runtime Logic
+- **caching** - cached outputs can be used per task (see [Caching[Caching]])
```
Broken link markdown

rimolive replied:
```suggestion
- **caching** - cached outputs can be used per task (see [Caching[Caching])
```

rimolive commented on content/en/docs/components/pipelines/concepts/pipeline.md:
```diff
@@ -5,20 +5,126 @@ weight = 10
+
+The KFP Python SDK client (`kfp.client.Client`) supports submitting pipeline runs to the KFP backend either directly (`create_run_from_pipeline_func`), or from the compiled pipeline YAML file (`create_run_from_pipeline_package`). The KFP UI also supports running pipelines once their YAML files have been uploaded (either from the GUI or Python client interface).
+
+The KFP documentation includes more detailed examples exploring pipeline implementation and runs. See the [Getting Started Guide][Getting Started] to quickly test out runing a pipeline on a KFP cluster. For a more in-depth treatment of pipeline implementation, visit the [User Guide][User Guide] section of the documentation, including sections on [Core Functions][Core Functions] and [Data Handling ...
```
Broken link markdown in the Core Functions

dandawg replied:
fixed

rimolive commented on content/en/docs/components/pipelines/concepts/pipeline-root.md:
```diff
@@ -5,60 +5,119 @@ weight = 15
-<img src="/docs/images/pipelines/v1/v2-compatible/pipelines-ui-pipelineroot.png"
-alt="Configure pipeline root on the pipelines UI"
-class="mt-3 mb-3 border rounded">
+A *pipeline root* represents a path within an object store bucket (MinIO, S3, GCS) where Kubeflow Pipelines stores [artifacts][artifact] from pipeline runs. Pipeline roots can be set at the cluster, [pipeline][pipeline], and [run][run] level, with support for authentication and overrides for specific paths provided.
```
broken markdown link for artifact

dandawg replied:
Fixed

rimolive commented on content/en/docs/components/pipelines/concepts/pipeline-root.md:
```diff
@@ -5,60 +5,119 @@ weight = 15
+A *pipeline root* represents a path within an object store bucket (MinIO, S3, GCS) where Kubeflow Pipelines stores [artifacts][artifact] from pipeline runs. Pipeline roots can be set at the cluster, [pipeline][pipeline], and [run][run] level, with support for authentication and overrides for specific paths provided.
+
+{{% alert title="Note" color="info" %}}
+It's important to understand how pipeline roots fit in KFP's data ecosystem. Pipeline roots are KFP's way to store [artifacts][artifact] (user data files) from [runs][run] in particular. Metadata on these [artifacts][artifact] (including their storage paths) is stored in an SQL database. Independently--and not to be confused with pipeline roots--KFP uses another object storage specification (in ` ...
```
all markdown links are broken

rimolive commented on content/en/docs/components/pipelines/user-guides/data-handling/pipeline-root.md:
```diff
@@ -0,0 +1,67 @@
+
+**references**:
+* [add-op-transformer](https://kubeflow-pipelines.readthedocs.io/en/stable/source/dsl.html#kfp.dsl.PipelineConf.add_op_transformer)
+<!-- TODO: add_op_transformer no longer exists in this doc reference -->
```
@HumairAK  Is this a SDK doc issue? looks like the reference does not exist anymore

dandawg replied:
Yeah, I believe this is outdated. I remember transformers from v1. This is the content that I moved from  the concept section to the the user-guides section. I didn't update it, just moved it.

An update to the doc is definitely needed, but that should probably be its own PR in my view. The IR YAML concept doc is in a similar boat (I just moved the content to a better place, but update will also be needed). When the content gets updated, the links can also be addressed. Does that work?

rimolive reviewed (changes requested):
I left some comments to fix the links

dandawg commented:
@rimolive I updated the links so that they work in the PR. I'm hoping the website will also know what to do with them. I did this for the new content I added (pipelines.md and pipeline-root.md in the concepts section).

The other content, is content that I moved (ir-yaml.md and the user-guide pipeline-root.md), but didn't change. This content definitely needs content updating (beyond the link issues), but I believe that is another PR.

hbelmiro commented on content/en/docs/components/pipelines/user-guides/data-handling/pipeline-root.md:
```diff
@@ -0,0 +1,67 @@
+
++++
+
+Starting from [Kubeflow Pipelines SDK v2](https://kubeflow-pipelines.readthedocs.io/en/stable/) and Kubeflow Pipelines v2, Kubeflow Pipelines supports a new intermediate artifact repository feature: pipeline root in both [standalone deployment](/docs/components/pipelines/operator-guides/installation/) and [AI Platform Pipelines](https://cloud.google.com/ai-platform/pipelines/docs).
```
I'm not sure we should mention products (AI Platform Pipelines).
Can't we just mention the Kubeflow Platform? Or not mention anything at all, if pipeline root is supported in both standalone and KF platform deployments?

```suggestion
Starting from [Kubeflow Pipelines SDK v2](https://kubeflow-pipelines.readthedocs.io/en/stable/) and Kubeflow Pipelines v2, Kubeflow Pipelines supports a new intermediate artifact repository feature: pipeline root.
```

dandawg replied:
I totally agree @hbelmiro . This is content I lifted from the previous doc, but I think your suggestion is an improvement. I'll add it in.

hbelmiro commented on content/en/docs/components/pipelines/user-guides/data-handling/pipeline-root.md:
```diff
@@ -0,0 +1,67 @@
+
+## Before you start
+This guide tells you the basic concepts of Kubeflow Pipelines pipeline root and how to use it.
+This guide assumes that you already have Kubeflow Pipelines installed, or want to use standalone or AI Platform Pipelines options in the [Kubeflow Pipelines deployment
+guide](/docs/components/pipelines/operator-guides/installation/) to deploy Kubeflow Pipelines.
```
Same as above regarding mentioning products.

dandawg replied:
Yeah, this was also from previous doc, but probably best that it's taken out.

dandawg replied:
Done.

hbelmiro commented on content/en/docs/components/pipelines/user-guides/data-handling/pipeline-root.md:
```diff
@@ -0,0 +1,67 @@
+Kubeflow Pipelines is configured with the authentication of the MinIO instance deployed with itself.
+
+#### GCS
+If you want to specify the `pipeline root` to GCS :
+
+check [authentication-pipelines](https://googlecloudplatform.github.io/kubeflow-gke-docs/docs/pipelines/authentication-pipelines/)
```
```suggestion
If you want to specify the `pipeline root` to GCS, check [authentication-pipelines](https://googlecloudplatform.github.io/kubeflow-gke-docs/docs/pipelines/authentication-pipelines/)
```

hbelmiro commented on content/en/docs/components/pipelines/user-guides/data-handling/pipeline-root.md:
```diff
@@ -0,0 +1,67 @@
+#### S3
+If you want to specify the `pipeline root` to S3, please choose one of the following options:
+
+* Via [AWS IRSA](https://aws.amazon.com/blogs/containers/cross-account-iam-roles-for-kubernetes-service-accounts/):
```
```suggestion
* Via [AWS IRSA](https://aws.amazon.com/blogs/containers/cross-account-iam-roles-for-kubernetes-service-accounts/)
```

hbelmiro commented on content/en/docs/components/pipelines/user-guides/data-handling/pipeline-root.md:
```diff
@@ -0,0 +1,67 @@
+Instructions can be found [here](/docs/components/pipelines/operator-guides/configure-object-store.md#kfp-launcher-object-store-configuration).
+
+####  Via Building Pipelines
+You can configure a pipeline root through the `kfp.dsl.pipeline` annotation when [building pipelines](/docs/components/pipelines/legacy-v1/sdk/build-pipeline/#build-your-pipeline)
```
```suggestion
You can configure a pipeline root through the `kfp.dsl.pipeline` annotation when [building pipelines](/docs/components/pipelines/legacy-v1/sdk/build-pipeline/#build-your-pipeline).
```

dandawg commented:
> Thank you. I left some comments.
Thanks @hbelmiro for helping find my typos. Great improvements!

dandawg commented:
@rimolive I ended up reverting the links to the original format. I believe this doesn't work in preview (or in markdown on the repo), but it works in the website. 

Existing links like this (/docs/components/pipelines/...) work on the website, and there is a developer note that suggests this pattern [here](https://github.com/kubeflow/website?tab=readme-ov-file#menu-structure):
> The left-hand navigation panel is defined by the directory structure under the [docs directory](https://github.com/kubeflow/website/tree/master/content/en/docs).

I have my complaints about the current setup. 
- The developer notes could be improved to add more understanding.
- It's hard to test links before they go live if they are broken in preview and markdown (ideally the website and these).
-  We should be able to test locally with hugo, but that also appears to be broken. I have a [bug opened](https://github.com/kubeflow/website/issues/4076).

hbelmiro reviewed (commented):
@dandawg I'm not sure I understand your comments about the broken links.
Anyway, I pointed some of them that are still broken/not rendered.

I see commits that don't look related to this PR. It seems like you did a merge instead of a rebase.

Also, don't forget to sign off your commits.

dandawg commented:
Just FYI, I'm busy preparing a Red Hat Summit talk. As soon as that is out of the way, I plan to fix this.

dandawg commented on content/en/docs/components/pipelines/concepts/pipeline-root.md:
```diff
@@ -5,60 +5,118 @@ weight = 15
+
+2. **KFP Pipeline Level**
+    
+    The KFP Python SDK also provides functionality for ML Engineers to set the `pipeline_root` parameter when creating a [pipeline][pipeline] via the `@dsl.pipeline` decorator. When set, this parameter overrides cluster-level default settings for the pipeline root when the [pipeline][pipeline] is run.
```
fixed

dandawg commented on content/en/docs/components/pipelines/concepts/pipeline.md:
```diff
@@ -5,20 +5,127 @@ weight = 10
+
+The KFP Python SDK client (`kfp.client.Client`) supports submitting pipeline runs to the KFP backend either directly (`create_run_from_pipeline_func`), or from the compiled pipeline YAML file (`create_run_from_pipeline_package`). The KFP UI also supports running pipelines once their YAML files have been uploaded (either from the GUI or Python client interface).
+
+The KFP documentation includes more detailed examples exploring pipeline implementation and runs. See the [Getting Started Guide][Getting Started] to quickly test out running a pipeline on a KFP cluster. For a more in-depth treatment of pipeline implementation, visit the [User Guide][User Guide] section of the documentation, including sections on [Core Functions][Core Functions] and [Data Handlin ...
```
fixed

dandawg commented:
@rimolive @hbelmiro thanks for catching the issues with links and the files that snuck in from merge. I was finally able to come back around and fix this. It is ready for your review now.

juliusvonkohout commented:
/lgtm
/approve

since it seems to significantly improve the status quo. @dandawg please create follow up PRs if needed.
